### PR TITLE
235 setup for deployment

### LINKED ISSUE #235: Setup UI and backend for deployment

### PR DESCRIPTION
**Pull request by kronken**
Closes #235 together with backend PR: https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/9

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,6 +3,12 @@ FROM node:10-alpine
 WORKDIR /build
 COPY . ./
 
+ARG BACKEND_API_URL
+ENV API_URL=${BACKEND_API_URL}
```
Comment (Divinare) [inline Dockerfile]: How about: BACKEND_API_URL=${BACKEND_API_URL} 

same with GEOSERVER_API_URL

File: README.md [hunk 2]
```diff
@@ -27,7 +27,12 @@ Runs at http://localhost:3000/
 ### Build and run in Docker container
 
 ```bash
+$ # Specify location of API:s
+$ docker build --build-arg BACKEND_API_URL=${API_URL} --build-arg BACKEND_GEOSERVER_URL=${GEOSERVER_URL} --tag=$DOCKER_IMAGE .
```
Comment (Divinare) [inline README.md]: Add an example of API_URL and GEOSERVER_URL?

File: src/components/map/Map.tsx [hunk 3]
```diff
@@ -90,6 +97,7 @@ class LeafletMap extends React.Component<IMapProps, IMapState> {
     }
 
     private fitBounds(bounds: L.LatLngBoundsExpression) {
+        this.getMap().invalidateSize();
```
Comment (Divinare) [inline src/components/map/Map.tsx]: Was this needed for something?

File: src/components/map/layers/NetworkLayers.tsx [hunk 4]
```diff
@@ -28,13 +22,15 @@ interface INetworkLayersProps {
     networkStore?: NetworkStore;
 }
 
+function getGeoServerUrl(layerName: string) {
+    const GEOSERVER_URL = process.env.GEOSERVER_URL || 'http://localhost:8080/geoserver';
+    // tslint:disable-next-line:max-line-length
+    return `${GEOSERVER_URL}/gwc/service/tms/1.0.0/joremapui%3A${layerName}@EPSG%3A900913@pbf/{z}/{x}/{y}.pbf`;
```
Comment (Divinare) [inline src/components/map/layers/NetworkLayers.tsx]: You forgot to include this very important comment :P
// TODO: Don't hardcode	
// Will be fixed in: https://github.com/HSLdevcom/jore-map-ui/issues/235

File: travis-build.sh [hunk 5]
```diff
@@ -5,12 +5,15 @@ ORG=${ORG:-hsldevcom}
 DOCKER_TAG=${TRAVIS_BUILD_NUMBER:-latest}
 DOCKER_IMAGE=$ORG/jore-map-ui:${DOCKER_TAG}
 DOCKER_IMAGE_LATEST=$ORG/jore-map-ui:latest
+DOCKER_IMAGE_DEV=$ORG/jore-map-ui:dev
```
Comment (Divinare) [inline travis-build.sh]: What is the difference between :dev and :latest?

File: typings.d.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+declare module '*.scss' {
+  interface IClassNames {
+    [className: string]: string
+  }
+  const classNames: IClassNames;
+  export = classNames;
+}
```
Comment (Divinare) [inline typings.d.ts]: Does this have to be in project root?

File: Dockerfile (hunk 1 above)
Comment (kronken) [reply Dockerfile]: I tried this, it doesn't work. I wanted it to work, It doesn't. I don't know why.

File: README.md (hunk 2 above)
Comment (kronken) [reply README.md]: I could add the dev servers url, true.

File: src/components/map/Map.tsx (hunk 3 above)
Comment (kronken) [reply src/components/map/Map.tsx]: Yes, I can add comment

File: src/components/map/layers/NetworkLayers.tsx (hunk 4 above)
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: Well, this is #235, I didn't really think it was hardcoded anymore. I guess it is somewhat hardcoded, but '/gwc/service/tms/1.0.0/joremapui%3A${layerName}@EPSG%3A900913@pbf' is just declaring the type of tiles we want. This functionality is connected to other things in this file, for example that we specify we read the tiles in tms format

File: travis-build.sh (hunk 5 above)
Comment (kronken) [reply travis-build.sh]: We usually have the practise of tagging images based on where they are used. :Latest is just a tag for the latest image.

File: typings.d.ts (hunk 6 above)
Comment (kronken) [reply typings.d.ts]: Maybe not, I don't know actually. The example in the documentation had it in the root, but we could try to move it if you have a better location for it

File: Dockerfile (hunk 1 above)
Comment (Divinare) [reply Dockerfile]: How about:
ENV BACKEND_API_URL=${_BACKEND_API_URL}?

File: README.md (hunk 2 above)
Comment (kronken) [reply README.md]: Fixed

File: src/components/map/Map.tsx (hunk 3 above)
Comment (kronken) [reply src/components/map/Map.tsx]: fixed

File: Dockerfile (hunk 1 above)
Comment (kronken) [reply Dockerfile]: These are the same variables that are used when building the docker image.
I don't think that the command: docker build --build-arg _BACKEND_API_URL="asd" .
complies to normal env variable standards

File: src/components/map/layers/NetworkLayers.tsx (hunk 4 above)
Comment (Divinare) [reply src/components/map/layers/NetworkLayers.tsx]: Ah yes right, I didnt notice this was #235 :P I agree it is less hard coded now. Part of it still is, but maybe it is not worth fixing to a separate file

File: typings.d.ts (hunk 6 above)
Comment (Divinare) [reply typings.d.ts]: The location depends on which file uses this? There is also images.d.ts, it should be moved too.

File: Dockerfile (hunk 1 above)
Comment (Divinare) [reply Dockerfile]: Didn't quite understand why API_URL=${BACKEND_API_URL} works but ENV BACKEND_API_URL=${_BACKEND_API_URL} doesn't (added an extra _ to make those variables different from each other).

File: typings.d.ts (hunk 6 above)
Comment (kronken) [reply typings.d.ts]: I can try to move them to a common place such as /typings/ or similar

Comment (Divinare) [reply typings.d.ts]: Okay, do that and my review is ok!

**Pull request merged**